sentry/src/main/java/io/sentry/GsonSerializer.java

```diff
@@ -169,13 +169,19 @@ public void serialize(
       writer.write("\n");
 
       for (final SentryEnvelopeItem item : envelope.getItems()) {
-        gson.toJson(item.getHeader(), SentryEnvelopeItemHeader.class, writer);
-        writer.write("\n");
-        writer.flush();
+        try {
```

Comment:
It's open up for the case of sending envelopes without items but if the stream in question is the network stream, it would be too late to do anything about it anyway unless we supported some trailing header to communicate the failure

Reply:
I think its already done but if not yet, it'd be nice to check if the envelope items are not empty before calling this method

Reply:
@marandaneto what exactly do you mean by checking if the envelope items are not empty?

Reply:
Good point @bruno-garcia, but I think we already had the problem before. I think this change slightly improves it. 


**Given an envelope with two items and the first to send throws an exception.**
_Old_
Send header
Send 1st item -> Exception
Stop

We just sent an envelope without items.

_New_
Send header
Send 1st item -> Exception, Drop the item
Send 2nd 

We send the envelope but dropped the 1st item.


**Given an envelope with one item that throws.**
_Old and New_
Send header
Send 1st item -> Exception

We just sent an envelope without items.


For output streams that are all or nothing we just changed the behavior, because if there is one error in any of the items we previously discarded the whole envelope and now we keep the envelope. Therefore we keep an envelope without any items. I think we could solve this by throwing an exception when we serialize an envelope and we can't serialize a single item.

Reply:
> @marandaneto what exactly do you mean by checking if the envelope items are not empty?

I just checked up and we don't do it:

so here:
https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/transport/AsyncConnection.java#L156

we could do:

```
if (!envelope.getItems().isEmpty()) {
      executor.submit(new EnvelopeSender(envelope, hint, currentEnvelopeCache));
}
```

ps: `getItems` is an iterator so I don't think `isEmpty` would work, just a pseudocode

so we'd never try to send an envelope without items.

Reply:
I think we already check that in https://github.com/getsentry/sentry-java/blob/ef1eef782fb0eeb2e1a0e5e7fbbb00fe253be167/sentry/src/main/java/io/sentry/transport/AsyncConnection.java#L143, so I think it is not needed.

Reply:
@philipphofmann this condition is only met if at least 1 item of the envelope is rate limited

Reply:
it's not related to this PR though, and the backend accepts an empty envelope, so feel free to ignore it, I just wrote it because I thought that the backend would trigger an error, it doesn't, as said by Armin.